### ISSUE TITLE
torch.compile: out= variants of CompositeImplicitAutograd linalg ops (linalg.norm, cholesky, ...) recompile on every input shape

### ISSUE THREAD

**Issue by jonasvq** [pytorch/pytorch#187953]
### Describe the bug

The `out=` variants of several `CompositeImplicitAutograd` linalg ops recompile
under `torch.compile` on every new input shape, while the functional variants
specialize only once. Reported originally as part of #135859 for `linalg.norm`
and `cholesky`; the `max` / `min` / `topk` cases (structured ops) are fixed in
#187952, but the composite ops need a different fix.

Minimal repro:

```python
import torch
torch._logging.set_logs(recompiles=True)

def f(x):
    out = x.new_empty(x.shape[0])
    torch.linalg.norm(x, dim=1, out=out)
    return out

cf = torch.compile(f)
for n in range(4, 10):
    cf(torch.randn(n, 6))   # recompiles on every n
```

`torch.linalg.norm(x, dim=1)` (no `out=`) does not recompile. `torch.cholesky`
/ `torch.linalg.cholesky` with `out=` behave the same way.

### Root cause (observed)

The composite `out=` kernel computes the functional result, then resizes the
user-provided out tensor, e.g. `linalg_norm_out` in
`aten/src/ATen/native/LinearAlgebra.cpp`:

```cpp
auto out = at::linalg_norm(...);              // functional: decomposes symbolically, fine
at::native::resize_output(result, out.sizes());   // specializes
result.copy_(out);
```

Under `FakeTensor`, `resize_output` dispatches `aten.size.default`, whose handler
materializes symbolic sizes to Python ints
(`torch/_subclasses/fake_tensor.py`, `_DISPATCH_META_HANDLERS`). Calling `int()`
on the `SymInt` adds an `Eq(s, 4)` guard and specializes the input dim, so the
frame recompiles whenever the shape changes.

The same `compute functional -> resize_output(result, tmp.sizes()) -> copy_`
pattern is used by many composite linalg `out=` kernels (cholesky, qr, eig, eigh,
svd, lstsq, ...), so they likely all share this behavior.

For structured ops (max/min/topk) the fix is to register a meta for the out
overload so FakeTensor never runs the C++ out= path (#187952). That approach is
not available here: registering a meta for a `CompositeImplicitAutograd` op is
explicitly disallowed (`activate_meta()` raises "we shouldn't register meta
function for it").

Possible directions (for maintainers to weigh):
- make `resize_output` / the `aten.size.default` fake handling symbolic-safe when
  the sizes already match, or
- trace these composite `out=` ops via their functional decomposition plus a
  symbolic out copy (the `out_wrapper` `_maybe_resize_out` path), instead of the
  C++ out= kernel.

### Error logs

```
[__recompiles] Recompiling function f ...
    triggered by the following guard failure(s):
    - 0/0: tensor 'x' size mismatch at index 0. expected 4, actual 5
```

### Versions

PyTorch built from source at `main` (2.14.0a0+git37c67b7), Python 3.12,
CUDA 12.8, Windows 11. Also reproduces on CPU.

---

Filed with the assistance of an AI coding assistant (Claude); reviewed by the author.

cc @jianyuh @nikitaved @mruberry @walterddr @Lezcano @chauhang @penguinwu @eellison @aorenste @ezyang @bobrenjc93 @aditvenk @laithsakka @oulgen @jamesjwu @anijain2305 @masnesral @coconutruben @bdhirsh

**Comment by Implycitt** [comment]
Hello,

I was wondering if I could take a look at this issue for a possible fix.

Thank you

**Comment by jonasvq** [comment]
Go ahead.

**Comment by jonasvq** [comment]
## Root-cause investigation: why both proposed directions don't work, and where the real fix lives

I dug into this with a local source build (added instrumentation + a rebuild) and want to record the findings, because **both candidate directions in the issue turn out to be insufficient**, for concrete and slightly surprising reasons. TL;DR: the specializing guard is added by the C++ `out=` kernel's `resize_output` *check* during Dynamo's fake-tensor propagation, and neither a Python decomposition nor a naive `sym_sizes()` swap removes it.

### Repro baseline (backend that only counts frames, `dynamic=True`)

```
linalg.norm    out=  -> 6 recompiles   (functional: 1)
linalg.cond    out=  -> 6 recompiles
linalg.vector_norm out= -> 1            (already fine)
```

`vector_norm` is fine because it is a *structured* op with a real meta kernel, so FakeTensor sets the output size directly and never calls `resize_output`. The CompositeImplicitAutograd ops (`linalg.norm`, `cond`, `pinv`, `tensorinv`, `tensorsolve`, plus the `BatchLinearAlgebra.cpp` family: `cholesky_inverse`, `geqrf`, `ormqr`, `linalg_eig`, `lstsq`, ...) all share the `compute tmp -> resize_output(out, tmp.sizes()) -> copy_` pattern.

### Direction 2 (register a Python decomposition for `aten.linalg_norm`) — does **not** fire

I registered the existing `torch._refs.linalg.norm` ref via `@register_decomposition(aten.linalg_norm)` (mirroring `vector_norm`). It is in `decomposition_table`, but instrumentation shows it is **called 0 times** during compile.

Reason: `FakeTensorMode.dispatch` only consults `decomposition_table` (fake_tensor.py ~L2978) for ops that actually reach the Python dispatch key. CompositeImplicitAutograd ops decompose in **C++ (Composite/Autograd key, above the Python key)** before they ever reach `FakeTensorMode.dispatch`, so the registered decomposition is bypassed. Only the leaf ops (`linalg_vector_norm`, `copy_`, `aten.size.default`) reach FakeTensor. This is why `vector_norm`'s ref "works" (it's a leaf with a meta) but `norm`'s ref never runs.

### Direction 1 (make the resize symbolic-safe) — necessary but the naive version is **insufficient**

The specializing guard `Eq(s, N)` comes from here, confirmed by an extended guard stack:

```
<C++ linalg_norm_out>  ->  out.sizes()
  -> aten.size.default
  -> _DISPATCH_META_HANDLERS[aten.size.default]: int(s) for s in args[0].size()   # fake_tensor.py ~L3622
  -> SymNode.guard_int  ->  eval Eq(s, 8) [guard added]
```

So the obvious step is to change the composite `out=` kernels from the `IntArrayRef` resize to the symint resize that several siblings in the same file already use (`LinearAlgebra.cpp:644/2081/2182`, `ReduceOps.cpp:1014`):

```cpp
- at::native::resize_output(result, out.sizes());
+ at::native::resize_output_symint(result, out.sym_sizes());
```

I applied this across the whole linalg family (26 sites in `LinearAlgebra.cpp` + `BatchLinearAlgebra.cpp`) and rebuilt. Result:

- In **isolation** (call the op under a fresh `FakeTensorMode` with symbolic inputs) the kernel is now provably clean: 0 guards, dispatches only `linalg_vector_norm` + `copy_`, the resize is a no-op.
- But under the **real Dynamo** `get_fake_value` path, `linalg.norm out=` **still recompiles 6x**. A single `run_node(target=linalg_norm)` still emits `aten.size.default -> int(s) -> Eq(s,8)`, and the dispatch order (`vector_norm` -> `size.default` x2 -> `copy_`) pins it to `resize_output_symint`'s **check**, i.e. `_resize_output_check_symint` -> `at::symint::sizes<SymInt>(output).equals(shape)` (`Resize.cpp:26`).

The reason the isolated probe looked clean but Dynamo doesn't: when `out` and the computed result carry the same symbol the `.equals` short-circuits, but in the real graph the equality check on `ArrayRef<SymInt>` is **not size-oblivious**, so it forces the symints to concrete ints. Contrast the Python `out_wrapper` path (`_maybe_resize_out` -> `_resize_output_check` -> `utils.same_shape`), which uses `guard_or_true(x != y)` and is exactly why structured/decomposed `out=` ops don't over-recompile.

### Where the real fix lives (maintainer decision)

A clean, localized fix doesn't exist; the options each have a design/blast-radius tradeoff:

1. **Make the C++ symint resize check size-oblivious** — give `_resize_output_check_symint` (`Resize.cpp`) `guard_or_true`/`sym_eq`-style semantics so a no-op resize on already-matching symbolic shapes doesn't specialize. This is the most direct analogue of the Python `same_shape` path, but it touches every `resize_output_symint` caller, so it needs a careful blast-radius review.
2. **Route CompositeImplicitAutograd `out=` ops through the symbolic-safe `out_wrapper` path during fake-tensor propagation** instead of running the C++ composite kernel — more localized to pt2 but a non-trivial dispatch change.

Option 1 looks like the smaller, more contained change and matches the existing `out_wrapper` precedent semantically. Happy to prototype it if maintainers agree on the direction.

(Investigation done with the assistance of an AI coding assistant; I reviewed and verified the findings on a local build.)
